### ISSUE TITLE
C2057 : Expected Constant Expression

### ISSUE THREAD

**Issue by jbltx** [mono/CppSharp#1079]
##### Brief Description
After compiling my generator for my native library, i launched it  and it generated a **Std-symbols.cpp** file, which obviously is needed to the generated CSharp files. I put it in a single DLL project and compile it, then the error **C2057** occured on the `std::basic_string` declaration.

Maybe I missed something about this file, but as far as i can see, the CSharp code needs a  `basic_string` entry point from a **Std-symbols.dll** file...

It seems to work fine in MSVC 2015 but not in  2017.

##### OS
Windows 10 x64

##### Used settings
```cs
GeneratorKind = GeneratorKind.CSharp;
// no pre/post-process
```
##### Target
MSVC 2017 for 64bits Platform

##### Stack trace or incompilable generated code
```cpp
#include <string>


template class __declspec(dllexport) std::char_traits<char>;
template class __declspec(dllexport) std::allocator<char>;
template class __declspec(dllexport) std::basic_string<char, std::char_traits<char>, std::allocator<char> >;
```

**Comment by tritao** [comment]
What is the error?

**Comment by jbltx** [comment]
@tritao  well, it's written in the title... I can't compile generated Std-symbols.cpp files using VS2017 because of the error `C2057 : Expected Constant Expression` which occurs on `basic_string` declaration.

To more, even if I am able to compile in VS2015, when I place the compiled **Std-symbols.dll** file next to my compiled CSharp executable, this error appears : 

```
Unhandled Exception: System.EntryPointNotFoundException: Unable to find an entry point named '?' in DLL 'Std-symbols'.
   at Std.BasicStringExtensions.__Internal.BasicString(IntPtr instance, String _Ptr, IntPtr _Al)
   at Std.BasicStringExtensions.BasicString(String _Ptr, Allocator`1 _Al) in F:\Git\MyProject\build\MyProject.CSharp\Release\Std.cs:line 332
   at MyProject.MyNamespace.MyClass.Open(String iFileNamePath) in F:\Git\MyProject\build\MyProject.CSharp\Release\MyClass.cs:line 687
   at ConsoleApplication1.Program.Main(String[] args) in F:\Git\MyProject\ConsoleApplication1\Program.cs:line 16
```
It seems to be a name mangling issue from the generated **Std.cs** file I guess. I was wondering how do you guess the mangled name which has currently inside generated before compiling the **Std-symbols.dll** ?

**Comment by jbltx** [comment]
Using [demumble](https://github.com/nico/demumble) I tested both entrypoints, mine and the one used in `Std.cs`

```bat
demumble.exe "??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@PEBDAEBV?$allocator@D@1@@Z"

>> public: __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >(char const * __ptr64,class std::allocator<char> const & __ptr64) __ptr64
```
```bat
demumble.exe "??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@QEBDAEBV?$allocator@D@1@@Z"
>> public: __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >(char const * __ptr64 const,class std::allocator<char> const & __ptr64) __ptr64
```
So in CppSharp they are using `char const * __ptr64 const` but I need to use `char const * __ptr64`.

**Comment by tritao** [comment]
Interesting, thanks for the help in tracking it down.

I am still not 100% sure why it is happening.

Maybe the signature of the standard library `basic_string` changed between VS2015 and VS2017?

Since you have both VS, could you check if that is the case?

Only thing I can think of is to make sure you are using the same VS version when setting up CppSharp to generate the bindings and when compiling the code.

You can force a VS `ParserOptions.SetupMSVC(VisualStudioVersion vsVersion)`.

**Comment by tritao** [comment]
Maybe that would also explain the `C2057 : Expected Constant Expression` error.

@ddobrev Do you have some time to check this one under VS 2017?

**Comment by ddobrev** [comment]
@tritao I'm afraid not but I think I know what the problem is. We most probably do not print some types correctly, that is, we forget the trailing "const". It's most probably somewhere in CppTypePrinter. @jbltx it would be great if you could build C++# from source and debug.

**Comment by jbltx** [comment]
@tritao @ddobrev thanks for your answers, indeed i could try to test in a debug build, the reason why i used the nuget package is because i didn't see any 0.8.19 version tag in Github... Do you have a specific commit to start with ? Or the current master is ok ?

Thanks.

**Comment by jbltx** [comment]
I also reported the bug for Visual Studio 2017 at Microsoft Issue Tracker too, and is fixed for next releases : 
https://developercommunity.visualstudio.com/content/problem/237125/c2057-expected-constant-expression-on-basic-string.html

I hope to find time this weekend to make further tests with a custom build from master, and see if it resolves everything.

**Comment by jbltx** [comment]
The problem was solved in VS 2017 15.7, it was a IDE's bug.

https://developercommunity.visualstudio.com/solutions/237468/view.html

Thank you.
